syntax = "proto3";

//CustomerService라는 gRPC 서비스를 정의
//GetAll: 매개변수 없이 CustomerList를 반환하는 메소드입니다.
//Get: CustomerRequestId를 매개변수로 받고, 해당 Customer를 반환하는 메소드입니다.
//Insert: Customer를 매개변수로 받고, 추가된 Customer를 반환하는 메소드입니다.
//Update: Customer를 매개변수로 받고, 업데이트된 Customer를 반환하는 메소드입니다.
//Remove: CustomerRequestId를 매개변수로 받고, 아무 값도 반환하지 않는 메소드입니다.

service CustomerService {
    rpc GetAll (Empty) returns (CustomerList) {}
    rpc Get (CustomerRequestId) returns (Customer) {}
    rpc Insert (Customer) returns (Customer) {}
    rpc Update (Customer) returns (Customer) {}
    rpc Remove (CustomerRequestId) returns (Empty) {}
}

message Empty {}

message Customer {
    string id = 1;
    string name = 2;
    int32 age = 3;
    string address = 4;
}

message CustomerList {
    repeated Customer customers = 1;
}

message CustomerRequestId {
    string id = 1;
}